src/lib/reasoners/sat_solver_util.ml

```diff
@@ -0,0 +1,222 @@
+
+  | Symbols.Name name, _ ->
+    let arg_tys = List.map Expr.type_info xs in
+    let typed_name = (name, arg_tys, ty) in
+    ModelMap.get_value typed_name xs mdl.Models.model
+  | _ -> None
+
+let (let*) = Option.bind
+
+let get_value (type a) (module SAT : S with type t = a) env l =
+  let* mdl = SAT.get_model env in
+  (* HOTFIX for SatML: we can assert new formula in [env.satml] only at the
+     level of decision [0]. After performing [unsat], the decision level isn't
+     necessary [0] and the SMT statement `get-value` can be called only in
+     SAT mode, so we have to reset all the decisions of the solver here.
+     Before resetting the decisions, we save the boolean model for sake
+     of optimization.
+
+     See issue: https://github.com/OCamlPro/alt-ergo/issues/1063 *)
+  let bmdl = SAT.get_boolean_model env in
+  SAT.reset_decisions env;
+  assert_model (module SAT) env mdl;
+  (* First, we attempt to retrieve each model term in the last computed
+     boolean and first-order models.
+
+     If we don't find the model term for an expression of [l], we assert a
+     new equation to force the solver to produce a model term for this
+     expression. *)
+  let l =
+    List.map
+      (fun e ->
+         match get_value_in_model mdl bmdl e with
+         | Some v -> Stdcompat.Either.Left v
+         | None ->
+           (* For each expression [e] of the list [l], we assert an equality
+              of the form [n = e] where [n] is a fresh name. *)
+           let ty = Expr.type_info e in
+           let id = Id.Namespace.GetValue.fresh () in
+           let sy = Symbols.name ~ns:GetValue id in
+           let name = Expr.mk_term sy [] ty in
+           let iff = match ty with Ty.Tbool -> true | _ -> false in
+           let t = Expr.mk_eq ~iff name e in
+           let () =
+             let { Expr.f; ty; _ } = Expr.term_view name in
+             match f with
+             | Symbols.Name name ->
+               SAT.declare env (name, [], ty)
+             | _ -> assert false
+           in
+           internal_assume (module SAT) env id t;
+           Stdcompat.Either.Right name
+      ) l
+  in
+  (* We have to check the satisfability of the new environment [env] in order
+     to produce a new model. If this call raise [Unsat], the model is wrong
+     and we cannot produce model terms for the expressions of [l]. *)
+  try
+    check (module SAT) env;
+    let* mdl = SAT.get_model env in
+    let bmdl = SAT.get_boolean_model env in
+    let values =
+      List.map
+        (fun v ->
+           match v with
+           | Stdcompat.Either.Left v -> v
+           | Stdcompat.Either.Right name ->
+             match get_value_in_model mdl bmdl name with
+             | Some v -> v
+             | None ->
+               (* The model generation has to produce a value for each
+                  declared names. If some declared names are missing in the
+                  model, it's a bug. *)
```

Comment:
If it's a plausible bug (I think it is) it would be better to use a an actual exception rather than an assertion failure.

Reply:
I agree. We shouldn't abort if the user wants to continue after this failure.

Reply:
I added a new exception `No_model` and caught it in `solving_loop` with an appropriate recoverable error message.